### DISCUSSION TITLE
Eglot choosing wrong server for a derived mode when there also exists a definition for the parent mode.

### DISCUSSION THREAD

**Discussion by Lenbok** [joaotavora/eglot#1460] [General]
I'm not sure how I am supposed to configure `eglot-server-programs` to avoid this problem.

I can add support for `groovy-mode` via:

```lisp
  (with-eval-after-load 'eglot
    (add-to-list 'eglot-server-programs
                 '((groovy-mode :language-id "groovy") . ("java" "-jar" "/home/len/dev/groovy-language-server/build/libs/groovy-language-server-all.jar"))))
```

But now I want to try out the language server for `nextflow-mode`, (which is derived from `groovy-mode`), using a similar construct:

```lisp
  (with-eval-after-load 'eglot
    (add-to-list 'eglot-server-programs
                 '((nextflow-mode :language-id "nextflow") . ("nextflow-language-server"))))
```

However, when I start eglot in a buffer that is in `nextflow-mode`, it starts the groovy language server instead.

**Comment by joaotavora** [comment]
I think it has to do with the order of entries in `eglot-server-programs`.  Show the contents of that variable. 

In theory, Eglot is doing nothing wrong.  `nextflow-mode` is telling it that "nextflow" is a dialect of "groovy" so starting the "groovy" server for "nextflow"  is not wrong under that lens.  But if it's not a dialect at all, then maybe `nextflow-mode` should not be deriving and rather use a different kind of mechanism to reuse some of `groovy-mode`'s code.   And then the order of entries in `e-s-programs` wouldn't matter.

**Comment by Lenbok** [comment]
OK, I found that by shuffling the order of the entries in `eglot-server-programs` I can get it to start the nextflow language server rather the groovy one.

I guess I was surprised by the assumption that because a mode X derives implementation from another mode Y, any language server for mode Y should be usable with mode X, when the server program list contains an entry that is a more specific match to the mode.

For now I'll change my groovy declaration to append the entry to the *end* of `eglot-server-programs` (and will do so with any servers for "generic" modes as yaml-mode that might have derived modes with their own servers).
